articles/react-native-firebase-text-recognition/index.md

```diff
@@ -0,0 +1,429 @@
+    <Text style={styles.buttonText}>Take Photo</Text>
+  </TouchableOpacity>
+  <TouchableOpacity style={styles.button} onPress={onSelectImagePress}>
+    <Text style={styles.buttonText}>Pick a Photo</Text>
+  </TouchableOpacity>
+<View>
+```
+
+Let's create a state to display the selected image on the UI.
+
+```JSX
+const [image, setImage] = useState();
+```
+
+Now, let's add an Image component below the buttons to display the selected image.
+
+```JSX
+<View>
+  <TouchableOpacity style={styles.button} onPress={onTakePhoto}>
+    <Text style={styles.buttonText}>Take Photo</Text>
+  </TouchableOpacity>
+  <TouchableOpacity style={styles.button} onPress={onSelectImagePress}>
+    <Text style={styles.buttonText}>Pick a Photo</Text>
+  </TouchableOpacity>
+  <Image source={{uri: image}} style={styles.image} resizeMode="contain" />
+</View>
+```
+
+Styles for the Image:
+
+```JSX
+image: {
+  height: 300,
+  width: 300,
+  marginTop: 30,
+  borderRadius: 10,
+},
+```
+
+Let's set the image state with the URI of the selected image in the `onImageSelect` function when the user did not cancel the operation.
+
+```JSX
+const onImageSelect = async (media) => {
+  if (!media.didCancel) {
+    setImage(media.uri);
+  }
+};
+```
+
+![Image UI](/engineering-education/react-native-text-recognition-with-firebase/with_image.jpg)
+
+### Recognize the text from the image
+
+Let's install the package for Firebase ML.
+
+```bash
+npm install @react-native-firebase/ml
+```
+
+Once the package is installed, let's import the package.
+
+```JSX
+import ml from '@react-native-firebase/ml';
+```
+
+We should use the `cloudDocumentTextRecognizerProcessImage` method in the `ml` package to process the image and recognize text from it.
+
+We will pass the URI of the selected image to this function.
+
+```JSX
+const result = await ml().cloudDocumentTextRecognizerProcessImage(media.uri);
+```
+
+The function will process the image and return the text recognized in the image along with an array of blocks of recognized text.
+
+Each [block](https://rnfirebase.io/reference/ml/mldocumenttextblock) will contain details about:
+
+- The bounding rectangle of the detected block of text in the image.
+
+- The confidence the maching learning service has in the result.
+
+- A list of recognized languages in that block.
+
+- An array of [paragaraphs](https://rnfirebase.io/reference/ml/mldocumenttextparagraph) recognized in the block of text.
+
+To learn more about the result object, refer to [the documentation](https://rnfirebase.io/reference/ml/mldocumenttext).
+
+
+Let's set up a state to store the result and render it in the UI.
+
+```JSX
+const [result, setResult] = useState({});
+```
+
+Let's set the state to the response of the `cloudDocumentTextRecognizerProcessImage` function.
+
+```JSX
+const onImageSelect = async (media) => {
+  if (!media.didCancel) {
+    setImage(media.uri);
+    const result = await ml().cloudDocumentTextRecognizerProcessImage(media.uri);
```

Comment:
I think this will bring an error since _const result_ is already initialized as a state variable.
Change the name of the variable and use the setResult function to set the state as the new variable.

Reply:
It won't. Whenever there is a Global variable and a Local variable are named the same, the Local variable will be used when it's referenced locally.

![image](https://user-images.githubusercontent.com/45763204/104831634-c48d1200-58b0-11eb-9ad0-e880832997e4.png)

Reply:
I agree, but that only works when you use the _let_ keyword in JS which means the variable can be reassigned.  You cannot add values to a constant variable in JS regardless of the function scope.
In our case we have const [result,setResult], we cannot reassign values to result, we can, however, use setResult to update the variable state. 
using _const result_ will therefore bring upon 2 errors. 

- result has already been declared,
- and a type error because you cannot assign to a constant variable

Reply:
I suggest you rename the result variable like this 

```javascript
const imageResult = await ml().cloudDocumentTextRecognizerProcessImage(media.uri);
    console.log(imageResult );
    setResult(imageResult);
```

Reply:
Have you looked at this? @zolomohan

Reply:
hmm, that makes perfect sense. But I'm not sure why this is working 😂. I tested the app and it worked fine without an error.

Reply:
Anyway, I think renaming the variable will also avoid any confusion like this.

Reply:
I followed the steps keenly and on my side there was an error.

Cheers for looking into it